README.md
# @mgsisk/postcss-modular-rhythm

PostCSS plugin that adds modular scale and vertical rhythm units to CSS.

[![Latest release][badge-release]][url-release]
[![Build status][badge-build]][url-build]
[![Code maintainability][badge-maintainability]][url-maintainability]
[![Test coverage][badge-coverage]][url-coverage]
[![Maintainer funding][badge-funding]][url-funding]

@mgsisk/postcss-modular-rhythm is a [PostCSS][] plugin that adds modular scale
and vertical rhythm units to CSS.

- `msu` – Modular Scale Unit, converted to a unitless value using the scale
  step, ratio, and bases.
- `mlh` – Modular Line Height, converted to a unitless value appropriate for
  `line-height` that will maintain vertical rhythm using the base line height
  and scale step, ratio, and bases.
- `mfs` – Modular Font Size, converted to a scaled value appropriate for `font`
  or `font-size` using the base font size and scale step, ratio, and bases.
- `vru` – Vertical Rhythm Unit, converted to a multiple of the base line height,
  appropriate for use with a variety of sizing properties (`height`, `margin`,
  `padding`, etc.)

## Installation

```sh
npm install --save-dev @mgsisk/postcss-modular-rhythm
```

## Usage

```js
// postcss.config.js

module.exports = {
  plugins: [
    …
    require('@mgsisk/postcss-modular-rhythm'),
    …
  ]
}
```

```css
/* input.css */

body {
  font: 0mfs/0mlh sans-serif;
}
 
h1 {
  font-size: 5mfs;
  line-height: 5mlh;
  margin: 0.5vru 1vru 2vru;
}
 
small {
  font: -1mfs/-1mlh serif;
}
```

```css
/* output.css */

body {
  font: 1em/1.5 sans-serif;
}
 
h1 {
  font-size: 2.48832em;
  line-height: 1.20563;
  margin: 0.75rem 1.5rem 3rem;
}
 
small {
  font: 0.83333em/1.8 sans-serif;
}
```

[Support resources][] are available if you need help with this project.

### Options

Set plugin options inside the `:root` pseudo-class:

```css
:root {
  --font-size: 12px;
  --line-height: 24px;
  --modular-scale: 1.68 1.1 1.4;
}
```

#### `--font-size`

Set the base font size and unit used for `mfs` values; defaults to `1em`. Mapped
to the `fontSize` and `fontUnit` properties in the JavaScript API.

#### `--line-height`

Set the base line height and vertical rhythm unit for `mlh` and `vru` values;
defaults to `1.5rem`. Mapped to the `lineHeight` and `rhythmUnit` properties in
the JavaScript API.

#### `--modular-scale`

Set the ratio and bases used to calculate `mfs`, `mlh`, and `msu` values;
defaults to `1.2 1`. The first number sets the scale ratio; following numbers
set optional bases. Use `msu` values to get the scale value for a given step in
the defined scale. For example, on the default scale - which [looks something
like this][] - `2msu` would produce a value of `1.44`. Mapped to the `ratio` and
`bases` properties in the JavaScript API.

#### `round`

You can set the number of decimal places to round to in the JavaScript API using
the `round` option; defaults to `5`.

```js
// postcss.config.js

module.exports = {
  plugins: [
    …
    require('@mgsisk/postcss-modular-rhythm')({round: 2}),
    …
  ]
}
```

### Vertical Rhythm

To maintain vertical rhythm, elements should have the same `mfs` and `mlh`
values. For example, an `h2` with `font-size: 4mfs;` should have `line-height:
4mlh;`.

## Contributing

[Contributions][] are always welcome; please read the [code of conduct][]
before you begin. See the [changelog][] for notable project changes, and report
any [security][] concerns you find.

## Thanks

To the [contributors][] that help to build, fund, and maintain this project;
the [other works][] that have contributed to and inspired this project; and
anyone that has found this project useful.

## License

[ISC][]

[badge-build]: https://img.shields.io/travis/com/mgsisk/postcss-modular-rhythm?link=https://travis-ci.com/mgsisk/postcss-modular-rhythm
[badge-coverage]: https://img.shields.io/codeclimate/coverage/mgsisk/postcss-modular-rhythm?link=https://codeclimate.com/github/mgsisk/postcss-modular-rhythm
[badge-funding]: https://img.shields.io/liberapay/receives/mgsisk?link=https://github.com/mgsisk/postcss-modular-rhythm/blob/master/docs/CONTRIBUTING.md%23funding
[badge-maintainability]: https://img.shields.io/codeclimate/maintainability/mgsisk/postcss-modular-rhythm?link=https://codeclimate.com/github/mgsisk/postcss-modular-rhythm
[badge-release]: https://img.shields.io/github/v/tag/mgsisk/postcss-modular-rhythm?sort=semver&link=https://github.com/mgsisk/postcss-modular-rhythm/releases
[changelog]: https://github.com/mgsisk/postcss-modular-rhythm/blob/master/docs/CHANGELOG.md
[code of conduct]: https://github.com/mgsisk/postcss-modular-rhythm/blob/master/docs/CODE_OF_CONDUCT.md
[contributions]: https://github.com/mgsisk/postcss-modular-rhythm/blob/master/docs/CONTRIBUTING.md
[contributors]: https://github.com/mgsisk/postcss-modular-rhythm/blob/master/docs/AUTHORS.md
[isc]: https://github.com/mgsisk/postcss-modular-rhythm/blob/master/LICENSE.md
[looks something like this]: https://www.modularscale.com/?1&em&1.2
[other works]: https://github.com/mgsisk/postcss-modular-rhythm/blob/master/docs/THANKS.md
[postcss]: https://postcss.org/
[security]: https://github.com/mgsisk/postcss-modular-rhythm/blob/master/docs/SECURITY.md
[support resources]: https://github.com/mgsisk/postcss-modular-rhythm/blob/master/docs/SUPPORT.md
[url-build]: https://travis-ci.com/mgsisk/postcss-modular-rhythm
[url-coverage]: https://codeclimate.com/github/mgsisk/postcss-modular-rhythm
[url-funding]: https://github.com/mgsisk/postcss-modular-rhythm/blob/master/docs/CONTRIBUTING.md#funding
[url-maintainability]: https://codeclimate.com/github/mgsisk/postcss-modular-rhythm
[url-release]: https://github.com/mgsisk/postcss-modular-rhythm/releases


changelog.md
# Changelog

## [0.2.2][] (2019-07-03)

### Changed

- Updated package dependencies

## [0.2.1][] (2019-03-19)

### Changed
- Replaced `jsonlint-cli` with `jsonlint`

### Fixed
- Inline documentation

## [0.2.0][] (2018-12-10)

### Fixed
- Plugin name

## [0.1.2][] (2018-12-09)

### Changed
- Replaced `js-yaml` with `yaml-lint`
- Replaced `jsonlint` with `jsonlint-cli`

## [0.1.1][] (2018-12-07)

### Changed
- Moved npm deployment to Travis configuration
- Package script names
- Replaced `markdownlint-cli` with `remark-cli`

## [0.1.0][] (2018-12-06)

> Initial public release.

### Added
- `mfs` unit support
- `mlh` unit support
- `msu` unit support
- `vru` unit support

[0.2.2]: https://github.com/mgsisk/eslint-config/compare/v0.2.1...v0.2.2
[0.2.1]: https://github.com/mgsisk/eslint-config/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/mgsisk/eslint-config/compare/v0.1.2...v0.2.0
[0.1.2]: https://github.com/mgsisk/eslint-config/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/mgsisk/eslint-config/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/mgsisk/postcss-modular-rhythm/tree/v0.1.0


tests/index.js
/* eslint-disable no-undefined */
const tap = require('tap')
const postcss = require('postcss')
const plugin = require('../src')

postcss([plugin])
  .process('a{width:calc(9px - 5msu)}', {from: undefined})
  .then((result)=> {
    tap.match(result.css, /9px - 2\.48832/u, 'It scales values up')
  })

postcss([plugin])
  .process('a{width:calc(9px - -5msu)}', {from: undefined})
  .then((result)=> {
    tap.match(result.css, /9px - 0\.40188/u, 'It scales values down')
  })

postcss([plugin])
  .process('a{font-size:0mfs}', {from: undefined})
  .then((result)=> {
    tap.match(result.css, /font-size:1em/u, 'It converts mfs values')
  })

postcss([plugin])
  .process('a{line-height:0mlh}', {from: undefined})
  .then((result)=> {
    tap.match(result.css, /line-height:1.5/u, 'It converts mlh values')
  })

postcss([plugin])
  .process('a{width:calc(9px - 0msu)}', {from: undefined})
  .then((result)=> {
    tap.match(result.css, /9px - 1/u, 'It converts msu values')
  })

postcss([plugin])
  .process('a{margin:1vru}', {from: undefined})
  .then((result)=> {
    tap.match(result.css, /margin:1.5rem/u, 'It converts vru values')
  })

postcss([plugin])
  .process('a{line-height:9mlh}', {from: undefined})
  .then((result)=> {
    tap.match(result.css, /line-height:1\.16284/u, 'It ensures a minimum line-height of 1')
  })

postcss([plugin])
  .process(':root{--font-size:12px}a{font-size:1mfs}', {from: undefined})
  .then((result)=> {
    tap.match(result.css, /font-size:14\.4px/u, 'It uses the correct font size unit')
  })

postcss([plugin])
  .process(':root{--line-height:16px}a{margin:1vru}', {from: undefined})
  .then((result)=> {
    tap.match(result.css, /margin:16px/u, 'It uses the correct rhythm unit')
  })

postcss([plugin])
  .process(':root{--modular-scale:1.1 2.2}a{width:calc(9px - 1msu)}', {from: undefined})
  .then((result)=> {
    tap.match(result.css, /9px - 2\.42/u, 'It uses custom modular scale values')
  })

postcss([plugin])
  .process(':root{--modular-scale:1.5}a{width:calc(9px - 0msu)}', {from: undefined})
  .then((result)=> {
    tap.match(result.css, /9px - 1/u, 'It uses a default base when none is provided')
  })

postcss([plugin])
  .process(':root{--bogus:1.5rem}a{margin:0vru}', {from: undefined})
  .then((result)=> {
    tap.match(result.css, /margin:0/u, 'It ignores unknown root parameters')
  })


src/index.js
const postcss = require('postcss')

/**
 * Calculate a modular scale value.
 *
 * @param {number} power The power of the modular scale value.
 * @param {number} ratio The modular scale ratio.
 * @param {Array} bases One or more modular scale bases.
 * @returns {number}
 */
function modularScale(power, ratio, bases) {
  const scale = []
  let step = 0

  while (Math.abs(step) <= Math.abs(power)) {
    for (let i = 0; i < bases.length; i++) {
      scale.push(bases[i] * Math.pow(ratio, step))
    }

    step += 1

    if (power < 0) {
      step -= 2
    }
  }

  return Array.from(new Set(scale))[Math.abs(step) - 1] // eslint-disable-line no-undef
}

/**
 * Calculate a unitless line height for a given modular scale.
 *
 * @param {number} lineHeight The base, unitless line height.
 * @param {number} power The power of the modular scale value.
 * @param {number} ratio The modular scale ratio.
 * @param {Array} bases One or more modular scale bases.
 * @returns {number}
 */
function lineHeightScale(lineHeight, power, ratio, bases) {
  const baseHeight = lineHeight / modularScale(power, ratio, bases)
  let realHeight = baseHeight

  while (realHeight < 1) {
    realHeight += baseHeight
  }

  return realHeight
}

/**
 * Convert custom modular scale and vertical rhythm units into valid CSS values.
 */
module.exports = postcss.plugin('@mgsisk/postcss-modular-rhythm', (options)=> {
  const defaults = {
    fontSize: 1,
    fontUnit: 'em',
    lineHeight: 1.5,
    rhythmUnit: 'rem',
    ratio: 1.2,
    bases: [1],
    round: 5,
  }

  options = {...defaults, ...options}

  return (root)=> {
    root.walkDecls((decl)=> {
      if (decl.parent.selector !== ':root') {
        return
      }

      if (decl.prop === '--font-size') {
        options.fontSize = parseFloat(decl.value)
        options.fontUnit = decl.value.replace(/(\d|\.|-)+/u, '')
      } else if (decl.prop === '--line-height') {
        options.lineHeight = parseFloat(decl.value)
        options.rhythmUnit = decl.value.replace(/(\d|\.|-)+/u, '')
      } else if (decl.prop === '--modular-scale') {
        const [ratio, ...bases] = postcss.list.space(decl.value)

        options.ratio = parseFloat(ratio)
        options.bases = Array.from(new Set(bases)).map(value=> parseFloat(value)) // eslint-disable-line no-undef

        if (!options.bases.length) {
          options.bases.push(1)
        }
      }
    })

    root.replaceValues(
      /(-?\d*\.?\d+)mfs\b/gu,
      (value)=> {
        const size = modularScale(parseFloat(value), options.ratio, options.bases) * options.fontSize

        return parseFloat(size.toFixed(options.round)) + options.fontUnit
      }
    )

    root.replaceValues(
      /(-?\d*\.?\d+)mlh\b/gu,
      (value)=> {
        const height = lineHeightScale(options.lineHeight, parseFloat(value), options.ratio, options.bases)

        return parseFloat(height.toFixed(options.round))
      }
    )

    root.replaceValues(
      /(-?\d*\.?\d+)msu\b/gu,
      value=> parseFloat(modularScale(parseFloat(value), options.ratio, options.bases).toFixed(options.round))
    )

    root.replaceValues(
      /(-?\d*\.?\d+)vru\b/gu,
      (value)=> {
        const rhythm = parseFloat(value) * options.lineHeight

        return parseFloat(rhythm.toFixed(options.round)) + options.rhythmUnit
      }
    )
  }
})
